libnethack/include/display.h

```diff
@@ -7,126 +8,168 @@
- */
-# define canseemon(mon) \
-    (mon->dlevel == level && \
-     (mon->wormno ? worm_known(mon) : \
-         (cansee(mon->mx, mon->my) || see_with_infrared(mon))) \
-     && mon_visible(mon))
+/* Astral vision. Requires that the target is either on a lit square or adjacent,
+   that the target is not invisible, hiding, or buried, and that the target is
+   within the source's astral vision range. */
+# define MSENSE_XRAY          0x00000010u
 
+/* Monster detection. Requires that the source has monster detection and that the
+   target is on the same level. */
+# define MSENSE_MONDETECT     0x00000020u
 
-/*
- * canspotmon(mon)
- *
- * This function checks whether you can either see a monster or sense it by
- * telepathy, and is what you usually call for monsters about which nothing is
- * known.
- */
-# define canspotmon(mon) (canseemon(mon) || sensemon(mon))
-
-/* knowninvisible(mon)
- * This one checks to see if you know a monster is both there and invisible.
- * 1) If you can see the monster and have see invisible, it is assumed the
- * monster is transparent, but visible in some manner.
- * 2) If you can't see the monster, but can see its location and you have
- * telepathy that works when you can see, you can tell that there is a
- * creature in an apparently empty spot.
- * Infravision is not relevant; we assume that invisible monsters are also
- * invisible to infravision.
- */
-# define knowninvisible(mon) \
-    (mtmp->minvis && \
-        ((cansee(mon->mx, mon->my) && (See_invisible || Detect_monsters)) || \
-        (!Blind && (HTelepat & ~INTRINSIC) && \
-            distu(mon->mx, mon->my) <= (BOLT_LIM * BOLT_LIM) \
-        ) \
-        ) \
-    )
+/* Warning vs. monster class, a property. Requires that the source has the
+   property and the target is on the same level and has the chosen monster
+   class. (At present, this is only granted by certain artifacts.) */
+# define MSENSE_WARNOFMON     0x00000040u
 
-/*
- * is_safepet(mon)
- *
- * A special case check used in attack() and domove().  Placing the
- * definition here is convenient.
- */
-# define is_safepet(mon) \
-        (mon && mon->mtame && canspotmon(mon) && flags.safe_dog \
-                && !Confusion && !Hallucination && !Stunned)
+/* Covetous sense. Requires that the source is covetous for an item that the
+   target is holding and the target is on the same level. */
+# define MSENSE_COVETOUS      0x00000080u
+
+/* Smell of gold. Implemented for xorns sensing the player in 3.4.3; the
+   behaviour is preserved and symmetrised for 4.3. */
+# define MSENSE_GOLDSMELL     0x00000100u
 
+/* Scent. Works over a radius-root-5 range, and requires that the source and
+   target are either both underwater or both out of water. The source must have
+   the M3_SCENT flag. TODO: make stinking clouds block this. */
+# define MSENSE_SCENT         0x00000200u
+
+/* Detect Fish. For the Trident of Fish */
+# define MSENSE_DETECTFISH    0x00000400u
```

Comment:
`MSENSE_DETECTFISH` was added for the Trident of Fish Command. It was also added to `MSENSE_ANYVISION` a couple lines down.